server/common/rest.py

```diff
@@ -114,48 +151,97 @@ def _query_parameter_to_filter(args):
     return result
 
 
-def schema_get_helper(data_adaptor):
+def schema_get_helper(data_adaptor, user_id=None):
     """helper function to gather the schema from the data source and annotations"""
-    schema = data_adaptor.get_schema()
+    schema = data_adaptor.get_schema(user_id=user_id)
     schema = copy.deepcopy(schema)
 
     return schema
 
 
-def genesets_get_helper(data_adaptor):
+def genesets_get_helper(data_adaptor, user_id=None):
     """helper function to get genesets present in the obs metadata"""
-    genesets = data_adaptor.get_genesets()
-    genesets = copy.deepcopy(genesets)
-    return genesets
+    payload = data_adaptor.get_genesets(user_id=user_id)
+    if isinstance(payload, dict):
+        result = copy.deepcopy(payload)
+        result.setdefault("genesets", [])
+        result.setdefault("tid", 0)
+        return result
+    genesets = copy.deepcopy(payload) if payload is not None else []
+    return {"genesets": genesets, "tid": 0}
 
 
 def schema_get(data_adaptor):
-    schema = schema_get_helper(data_adaptor)
+    user_id = _resolve_request_user_id(request)
+    schema = schema_get_helper(data_adaptor, user_id=user_id)
     return make_response(jsonify({"schema": schema}), HTTPStatus.OK)
 
 
 def genesets_get(data_adaptor):
```

Comment:
Genesets GET now flows through `_resolve_request_user_id` so returning a user-specific payload (genes + tid) becomes possible. Older adaptors that only expose `get_gene_sets` still work, but TileDB-backed paths return `{ genesets, tid }` to keep autosave in sync.